src/ast/passes/codegen_llvm.cpp

```diff
@@ -3054,81 +3061,103 @@ Value *CodegenLLVM::createLogicalOr(Binop &binop)
-  // log2() returns a bucket index for the given value. Index 0 is for
-  // values less than 0, index 1 is for 0, and indexes 2 onwards is the
-  // power-of-2 histogram index.
+  // Arguments: N (int64), K (0..5)
+  // Maps each power of 2 into 2^K buckets, so we can build fine-grained
+  // histograms with low runtime cost.
+  //
+  // Returns 0 for N < 0, 1..2^K for 0 <= N < 2^K; otherwise
+  // N has the form ..01xxx... where the leftmost is at bit L >= K,
+  // xxx are the following K bits with a value 0 <= X < 2^K.
+  // In this case the function returns 1 + (L-K + 1) * 2^K + X
   //
-  // log2(int n)
+  // log2(int n, int k)
   // {
-  //   int result = 0;
-  //   int shift;
-  //   if (n < 0) return result;
-  //   result++;
-  //   if (n == 0) return result;
-  //   result++;
-  //   for (int i = 5; i >= 0; i--)
-  //   {
-  //     shift = (n >= (1<<(1<<i))) << i;
+  //   if (n < 0) return 0;
+  //   mask = (1ul << k) - 1;
+  //   if (n <= mask) return n + 1;
+  //   n0 = n;
+  //   // find leftmost 1
+  //   l = 0;
+  //   for (int i = 5; i >= 0; i--) {
+  //     threshold = 1ul << (1<<i)
+  //     shift = (n >= threshold) << i;
   //     n >>= shift;
-  //     result += shift;
+  //     l += shift;
   //   }
-  //   return result;
+  //   l -= k;
+  //   // mask K bits after leftmost 1
+  //   x = (n0 >> l) & mask;
+  //   return ((l + 1) << k) + x;
   // }
 
-  FunctionType *log2_func_type = FunctionType::get(b_.getInt64Ty(), {b_.getInt64Ty()}, false);
+  FunctionType *log2_func_type = FunctionType::get(b_.getInt64Ty(), {b_.getInt64Ty(), b_.getInt64Ty()}, false);
   Function *log2_func = Function::Create(log2_func_type, Function::InternalLinkage, "log2", module_.get());
   log2_func->addFnAttr(Attribute::AlwaysInline);
   log2_func->setSection("helpers");
   BasicBlock *entry = BasicBlock::Create(module_->getContext(), "entry", log2_func);
   b_.SetInsertPoint(entry);
 
-  // setup n and result registers
-  Value *arg = log2_func->arg_begin();
+  // storage for arguments
   Value *n_alloc = b_.CreateAllocaBPF(CreateUInt64());
-  b_.CreateStore(arg, n_alloc);
-  Value *result = b_.CreateAllocaBPF(CreateUInt64());
-  b_.CreateStore(b_.getInt64(0), result);
+  b_.CreateStore(log2_func->arg_begin(), n_alloc);
+  Value *k_alloc = b_.CreateAllocaBPF(CreateUInt64());
+  b_.CreateStore(log2_func->arg_begin() + 1, k_alloc);
 
   // test for less than zero
   BasicBlock *is_less_than_zero = BasicBlock::Create(module_->getContext(), "hist.is_less_than_zero", log2_func);
   BasicBlock *is_not_less_than_zero = BasicBlock::Create(module_->getContext(), "hist.is_not_less_than_zero", log2_func);
-  b_.CreateCondBr(b_.CreateICmpSLT(b_.CreateLoad(b_.getInt64Ty(), n_alloc),
-                                   b_.getInt64(0)),
-                  is_less_than_zero,
-                  is_not_less_than_zero);
+
+  // code generation helper macros: Constant, LoaD
+#define _C(_n)		b_.getInt64(_n)
+#define _LD(_n)		b_.CreateLoad(b_.getInt64Ty(), _n)
```

Comment:
Nit: for consistency with rest of codebase, please avoid these macros

Reply:
ack, will do in next revision

Reply:
done